### DISCUSSION TITLE
Seeking clarification for HDT with Sparql / nquads

### DISCUSSION THREAD

**Discussion by C-Loftus** [comunica/comunica#1429] [Q&A] [answered]
Hello, I am working on a linked data project with a fair amount of data ~22million statements and am looking to use HDT to make queries more efficient.

My linked data is currently expressed as nquads inside graphdb, but from what I can see, nquads can't be directly encoded as HDT directly but needs HDTQ which is less supported generally. I just wanted to confirm this understanding and clarify if comunica or any other tool you are aware of supports nquads as HDT(Q) queryable with sparql

As I have been looking through options for this task, I also wanted to clarify tradeoffs between different options
- I saw [server.js](https://github.com/LinkedDataFragments/Server.js) and it seems that there are shared contributors on this project, but I am unclear if comunica is strictly superior or if there are any tradeoffs. 
    - On https://linkeddatafragments.org/software/ I see that comunica is listed as a client but this is a bit confusing to me since I thought you can serve HDT at an endpoint, i.e. like `comunica-sparql-hdt-http hdt@datasets/dataset-100M.hdt`
- For context, I don't need federation and would prefer if I can serve a sparql endpoint for static, immutable HDT file without needing to expose an associated graph database as well. I believe I can do that with comunica but wasn't sure

**Comment by rubensworks** [comment, marked as answer]
Hi @C-Loftus 👋,

Comunica makes use of the [C++ version of HDT](https://github.com/rdfhdt/hdt-cpp), which has no support for quads indeed. But if someone were to add support, Comunica will also be able to make use of it quite easily.

Server.js also makes use of the C++ version, so this one won't support HDTQ either. Server.js only focuses on exposing datasets using the [TPF interface](https://linkeddatafragments.org/specification/triple-pattern-fragments/). Comunica supports querying over such TPF interfaces, which is why it's listed there as a client. But you're right that Comunica can also be used to expose a (SPARQL endpoint) server interface over HDT files.

One additional option you could consider trying is just loading up everything in-memory from a plain N-Quads file, using `comunica-sparql-file-hdt`, since 22M triples may fit into memory. You'll probably have to increase the max memory usage of Node.js. I'm not sure how much memory you'll need, as this will depend on what your dataset looks like. Performance-wise, execution be a bit slower compared to HDT (10~20% based on our [last experiments](https://github.com/comunica/comunica-performance-results)), but that's better than not being able to query at all :-) Startup will also be slower, since indexing needs to happen at startup time (or the first time a query is executed).

**Reply by rubensworks** [reply]
Another option would be to add an actor to Comunica that adds bindings to HDT-Java. But I don't know atm how easy and performant this would be.

**Reply by C-Loftus** [reply]
Thanks very much for your response @rubensworks! That is super helpful. 

> loading up everything in-memory from a plain N-Quads file, using comunica-sparql-file-hdt, 

Is `@comunica/query-sparql-hdt` equivalent to `comunica-sparql-file-hdt`? I don't see the latter. And I think I am a bit confused, why would the `hdt` library allow loading in a plain N-Quads file that isn't HDT? 

Just curious, does the [java library](https://github.com/rdfhdt/hdt-java) support quads? I don't see any indication in the source code. I saw there was an [HDTQ fork](https://github.com/Rummy23/hdtq-java) for that java library, but it doesn't seem maintained and was mainly for research purposes.

**Reply by rubensworks** [reply]
> Is @comunica/query-sparql-hdt equivalent to comunica-sparql-file-hdt? I don't see the latter. And I think I am a bit confused, why would the hdt library allow loading in a plain N-Quads file that isn't HDT?

Oh, apologies. I meant [`comunica-sparql-file`](https://www.npmjs.com/package/@comunica/query-sparql-file).

> Just curious, does the [java library](https://github.com/rdfhdt/hdt-java) support quads? I don't see any indication in the source code. I saw there was an [HDTQ fork](https://github.com/Rummy23/hdtq-java) for that java library, but it doesn't seem maintained and was mainly for research purposes.

The fork would have to be used for getting quads support indeed. I thought it was meant to get merged into the upstream repo, but I am not aware of what the status on that is.

**Reply by C-Loftus** [reply]
Makes sense! Thanks! From what I can tell the HDTQ support in the Java lib isn't merged upstream. I think I will try the in-memory option you say and go from there!
